### PR TITLE
Add empty() function to Sink

### PR DESCRIPTION
**Pull request by Xaeroxe**
This addresses https://github.com/tomaka/rodio/issues/104

### CODE REVIEW

File: src/sink.rs [hunk 1]
```diff
@@ -40,13 +42,14 @@ impl Sink {
...
             stopped: Arc::new(AtomicBool::new(false)),
+            empty: Arc::new(AtomicBool::new(true)),
             detached: false,
         }
     }
 
     /// Appends a sound to the queue of sounds to play.
     #[inline]
-    pub fn append<S>(&self, source: S)
+    pub fn append<S>(&mut self, source: S)
```
Comment (tomaka) [inline src/sink.rs]: `append` really should keep `&self` instead of `&mut self`

File: src/sink.rs [hunk 2]
```diff
@@ -70,7 +73,8 @@ impl Sink {
                 }
             })
             .convert_samples();
-
+        self.empty = Arc::new(AtomicBool::new(false));
```
Comment (tomaka) [inline src/sink.rs]: Why not do `self.empty.store(false, Ordering::Relaxed);` instead?

File: src/spatial_sink.rs [hunk 3]
```diff
@@ -53,7 +53,7 @@ impl SpatialSink {
 
     /// Appends a sound to the queue of sounds to play.
     #[inline]
-    pub fn append<S>(&self, source: S)
+    pub fn append<S>(&mut self, source: S)
```
Comment (tomaka) [inline src/spatial_sink.rs]: Same here. Revert to `&self`.

**Comment by Xaeroxe** [comment]
So I'm going to address all your review comments here:

The idea behind this PR is that when a source returns `None` from `next()` that it stores "true" in the arc bool it contains.  So every source append must generate a new `Arc<AtomicBool>` and update the sink to point to that new `Arc<AtomicBool>` as a result during the append operation we require mutability.  Simply storing `false` in the existing `Arc<AtomicBool>` will break the check for when the most recently added source ends.

**Comment by tomaka** [comment]
Ah yeah that's true.
However another solution could be to use an `AtomicUsize`. When a sound is added, the counter is incremented. When a sound is finished, it is decremented. When the counter is equal to 0, the sink is empty.

**Comment by Xaeroxe** [comment]
That's much better.  Thanks, I'll do that.

**Comment by Xaeroxe** [comment]
@tomaka Ready for your review.

**Pull request merged**